### DISCUSSION TITLE
How to properly retrieve files from spdlog?

### DISCUSSION THREAD

**Discussion by maxim-lbmv** [gabime/spdlog#3065] [Q&A]
I need to send log files to the server. After successful transfer of logs to the server, local log files must be deleted. To write logs to files, I use rotating_file_sink_mt. What is the correct way to retrieve current logging data from spdlog for transmission to the server?

**Comment by tt4g** [comment]
You can use `spdlog::file_event_handlers`:

https://github.com/gabime/spdlog/blob/a2b4262090fd3f005c2315dcb5be2f0f1774a005/include/spdlog/common.h#L328-L339

For example, `spdlog::file_event_handlers::after_close(const spdlog::filename_t &)' is a callback that is called after the log file is closed.

https://github.com/gabime/spdlog/blob/a2b4262090fd3f005c2315dcb5be2f0f1774a005/include/spdlog/details/file_helper-inl.h#L85-L98

**Reply by maxim-lbmv** [reply]
> For example, `spdlog::file_event_handlers::after_close(const spdlog::filename_t &)' is a callback that is called after the log file is closed.

For rotating_file_sink, the close function is called only when the file size limit is reached and it is renamed, that is, until this moment it will be impossible to send logs, and I need to send them upon request from the server. Are there other ways?

**Reply by tt4g** [reply]
> I need to send them upon request from the server. Are there other ways?

spdlog does not have the features to accept requests and send logs.
That API should be created by you.

It is not clear what "the correct way to retrieve current logging" means, but if it is the data of the current log file, you can get the name of the current log file from `spdlog::rotating_file_sink::filename()`, so you can use ` fopen()` to access the file and send the data to the server.

https://github.com/gabime/spdlog/blob/a2b4262090fd3f005c2315dcb5be2f0f1774a005/include/spdlog/sinks/rotating_file_sink-inl.h#L66-L70

**Reply by maxim-lbmv** [reply]
Is it possible to make the rotate_ function available to the user to make things easier for me? Then, if I had two log files, I would move log_file_name.1 to the server, then call the rotate_ function and move log_file_name.1 to the server again.

**Reply by tt4g** [reply]
That should be accomplished by `spdlog::file_event_handlers`.
If you really want to implement the rotation operation with custom logic, please create your own custom sink: https://github.com/gabime/spdlog/wiki/4.-Sinks#implementing-your-own-sink
